### ISSUE TITLE
Agrupamiento (Clustering - KMEANS) - Cesar R. Mamani Turpo

### ISSUE THREAD

**Issue by rasec-420** [frdtorres/mainf24-2-BIDA#9]
from flask import Flask,jsonify, request
from flask_sqlalchemy import SQLAlchemy
from flask_marshmallow import Marshmallow
from sklearn.cluster import KMeans
from sklearn.preprocessing import LabelEncoder, StandardScaler
import pandas as pd
from datetime import datetime
import matplotlib
matplotlib.use('Agg')  
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import numpy as np

app = Flask(__name__)

app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://camarapu_datacovid:d4t4c0vid@192.185.48.151:3306/camarapu_datacovid'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)
ma = Marshmallow(app)

class Fallecido(db.Model):
    __tablename__ = 'fallecidos_covid'
    uuid = db.Column(db.Integer, primary_key=True)
    fecha_fallecimiento = db.Column(db.Integer)
    edad_declarada = db.Column(db.Integer)
    sexo = db.Column(db.String(9))
    fecha_nacimiento = db.Column(db.String(8))
    departamento = db.Column(db.String(13))
    provincia = db.Column(db.String(25))
    distrito = db.Column(db.String(36))

class FallecidoSchema(ma.Schema):
        class Meta:
                fields = ('uuid','fecha_fallecimiento','edad_declarada','sexo','fecha_nacimiento','departamento','provincia','distrito')

fallecido_schema = FallecidoSchema()
fallecidos_schema = FallecidoSchema(many=True)

# Ruta para realizar agrupamiento K-Means por edad y región
@app.route('/agrupamiento', methods=['GET'])
def realizar_agrupamiento():
    all_fallecidos = Fallecido.query.all()
    # Convertir los datos a un DataFrame de pandas
    fallecidos_data = fallecidos_schema.dump(all_fallecidos)
    df = pd.DataFrame(fallecidos_data)

    # Filtrar registros con fechas nulas o vacías
    df = df[(df['fecha_nacimiento'].notnull()) & (df['fecha_fallecimiento'].notnull()) & 
            (df['fecha_nacimiento'] != '') & (df['fecha_fallecimiento'] != '')]

    # Comprobar si hay registros 
    if df.empty:
        return jsonify({"error": "No hay registros válidos para procesar."}), 400

    # Convertir las fechas de nacimiento y fallecimiento a edades
    def calcular_edad(fecha_nac, fecha_fallecimiento):
        try:
            fecha_nac = datetime.strptime(fecha_nac, '%Y%m%d')
            fecha_fallecimiento = datetime.strptime(fecha_fallecimiento, '%Y%m%d')
            return fecha_fallecimiento.year - fecha_nac.year - ((fecha_fallecimiento.month, fecha_fallecimiento.day) < (fecha_nac.month, fecha_nac.day))
        except ValueError:
            return None  # Retornar None si hay un error en la conversión

    df['edad'] = df.apply(lambda row: calcular_edad(row['fecha_nacimiento'], row['fecha_fallecimiento']), axis=1)

    # Filtrar registros donde la edad sea válida
    df = df[df['edad'].notnull()]

    if df.empty:
        return jsonify({"error": "No se pudieron calcular edades válidas."}), 400

    # Codificar las columnas categóricas (departamento, provincia, distrito)
    le = LabelEncoder()
    df['departamento_encoded'] = le.fit_transform(df['departamento'])
    df['provincia_encoded'] = le.fit_transform(df['provincia'])
    df['distrito_encoded'] = le.fit_transform(df['distrito'])

    # Seleccionar las columnas necesarias para el agrupamiento
    X = df[['edad', 'departamento_encoded', 'provincia_encoded', 'distrito_encoded']]

    # Estandarización de los datos
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)

    # Aplicar K-Means con 5 clusters
    kmeans = KMeans(n_clusters=5)
    df['cluster'] = kmeans.fit_predict(X_scaled)

    # Devolver los resultados del agrupamiento
    resultado = df[['uuid', 'edad', 'departamento', 'provincia', 'distrito', 'cluster']].to_dict(orient='records')

    return jsonify(resultado)

##FUNCION CALCULAR EDAD
def calcular_edad(fecha_nac, fecha_fallecimiento):
    try:
        fecha_nac = datetime.strptime(fecha_nac, '%Y%m%d')
        fecha_fallecimiento = datetime.strptime(fecha_fallecimiento, '%Y%m%d')
        return fecha_fallecimiento.year - fecha_nac.year - ((fecha_fallecimiento.month, fecha_fallecimiento.day) < (fecha_nac.month, fecha_nac.day))
    except ValueError:
        return None
    

##Graficar CLUSTERS
@app.route('/graficar_clusters', methods=['GET'])
def graficar_clusters():
    all_fallecidos = Fallecido.query.all()
    fallecidos_data = fallecidos_schema.dump(all_fallecidos)
    df = pd.DataFrame(fallecidos_data)

    df = df[(df['fecha_nacimiento'].notnull()) & (df['fecha_fallecimiento'].notnull()) & 
            (df['fecha_nacimiento'] != '') & (df['fecha_fallecimiento'] != '')]

    if df.empty:
        return jsonify({"error": "No hay registros válidos para graficar."}), 400
    
    # Calcular edades
    df['edad'] = df.apply(lambda row: calcular_edad(row['fecha_nacimiento'], row['fecha_fallecimiento']), axis=1)
    df = df[df['edad'].notnull()]

    if df.empty:
        return jsonify({"error": "No se pudieron calcular edades válidas para graficar."}), 400

    # Codificación de las columnas categóricas
    df['departamento_encoded'] = df['departamento'].astype('category').cat.codes
    df['provincia_encoded'] = df['provincia'].astype('category').cat.codes
    df['distrito_encoded'] = df['distrito'].astype('category').cat.codes

    # Estandarización y K-Means
    X = df[['edad', 'departamento_encoded', 'provincia_encoded', 'distrito_encoded']]
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)

    kmeans = KMeans(n_clusters=5)
    df['cluster'] = kmeans.fit_predict(X_scaled)

    # Reducir dimensionalidad a 2D usando PCA
    pca = PCA(n_components=2)
    X_pca = pca.fit_transform(X_scaled)

    # Agregar las coordenadas PCA al DataFrame
    df['PCA1'] = X_pca[:, 0]
    df['PCA2'] = X_pca[:, 1]

    # Graficar los clusters
    plt.figure(figsize=(10, 6))
    sns.scatterplot(x='PCA1', y='PCA2', hue='cluster', data=df, palette='viridis', alpha=0.7)

    plt.title('Clusters de Fallecidos (K-Means)')
    plt.xlabel('Componente Principal 1')
    plt.ylabel('Componente Principal 2')
    plt.legend(title='Cluster')
    plt.grid(True)

    # Guardar la gráfica en un archivo y devolver la ruta
    plt.savefig('clusters.png')
    plt.close()  # Cerrar la figura para liberar memoria

    return jsonify({"message": "Gráfica generada exitosamente.", "file": "clusters.png"})

@app.route('/',methods=['GET'])
def index():
    return jsonify({'Mensaje':'Bienvenido'})

if __name__ == "__main__":
    app.run(debug=True)
